contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,329 @@ const (
+us-central1-b".
+
+A StorageClass update controller can be provided to convert existing topology
+restrictions specified under plugin-specific parameters to the new AllowedTopology
+field.
+
+The AllowedTopology will also be provided to provisioners as a new field, detailed in
+the provisioner section.  Provisioners can use the allowed topology information
+in the following scenarios:
+* StorageClass is using the default immediate binding mode. This is the
+  legacy topology-unaware behavior. In this scenario, the volume could be
+  provisioned in a domain that cannot run the Pod since it doesn't take any
+  scheduler input.
+* For volumes that span multiple domains, the AllowedTopology can restrict those
+  additional domains.  However, special care must be taken to avoid specifying
+  conflicting topology constraints in the Pod. For example, the administrator could
+  restrict a multi-zonal volume to zones 'zone1' and 'zone2', but the Pod could have
+  constraints that restrict it to 'zone1' and 'zone3'.  If 'zone1'
+  fails, the Pod cannot be scheduled to the intended failover zone.
+
+Kubernetes will leave validation and enforcement of the AllowedTopology content up
+to the provisioner.
+
+TODO: would this be better represented as part of StorageClass quota since that
+is where admins currently specify policies?
+However, ResourceQuota currently only supports quantity, and topology
+restrictions have to be evaluated during scheduling, not admission.
+
+#### Zonal Example
+This example restricts the volumes provisioned to zones us-central1-a and
+us-central1-b.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: zonal-class
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-standard
+allowedTopology:
+  required:
+    failure-domain.beta.kubernetes.io/zone:
+      values:
+      - us-central1-a
+      - us-central1-b
+```
+
+#### Multi-Zonal Example
+This example restricts the volume's primary and failover zones
+to us-central1-a and us-central1-b. Topologies that are incompatible with the
+storage provider parameters will be enforced by the provisioner.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: multi-zonal-class
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-standard
+  replicas: 2
+allowedTopology:
+  required:
+    failure-domain.beta.kubernetes.io/zone:
+      values:
+      - us-central1-a
+      - us-central1-b
+```
+
+### Storage Provider Capacity Reporting
+Some volume types have capacity limitations per topology domain. For example,
+local volumes have limited capacity per node. Provisioners for these volume
+types can report these capacity limitations through a new StorageClass status
+field.
+
+```
+type StorageClass struct {
+    ...
+
+    Status *StorageClassStatus
+}
+
+type StorageClassStatus struct {
+    Capacity *StorageClassCapacity
+}
+
+type StorageClassCapacity struct {
+    TopologyKey string
```

Comment:
Still I'm concerned about the Topology key in StorageClassCapacity, if we place it in the status field, and leave it to provisioners to specify, then for edge-case that:
 
 When scheduler find a storage with StorageClassCapacity empty, it is hard to tell whether the storage is indeed without capacity limit, or it's just of a state that capacity has not been initialized by provisioners yet.

 If we move the Topology key out of status field, and leave it to admin, it is thought to be fulfilled on StorageClass creation.
 And the edge-case above could be avoid.

Reply:
Then it will open up a new mode for failure, where administrator specifies the wrong topologyKey and the provisioner values won't match with it.

If a pod gets scheduled before provisioners have reported their capacity, then:
* Provisioning could fail.  In that case we should retry scheduling
* Provisioning could succeed.  Then we will not have marked the PVC with provisioned capacity information, so our calculations for used capacity will be incorrect.  Is there a way we can handle this?  Ie, have a StorageClass modify handler, when StorageClassCapacity is initialized, then initialize the capacity cache?

Reply:
Oh, I would prefer the scheduler could recognize the StorageClass at the first beginning, but not manage things by trying Provisioning. That makes things complex and may have effect on performance.

If you are worried that admin may specify a wrong key, then a simple flag to imply whether the StorageClass is of  capacity limit may also work.
I think the admin should at least know the StorageClass he creates is of  capacity limit or not.

Reply:
How will the scheduler know that all the provisioners have initialized and reported capacity? (in the case of local volume provisioners)

Reply:
Nm, in that case, once the topologyKey is specified, then capacity is zero until otherwise updated.

Reply:
> How will the scheduler know that all the provisioners have initialized and reported capacity? (in the case of local volume provisioners)

I think it doesn't need to know, what the scheduler cares:
1. Whether the StorageClass is of capacity limit
2. If so, whether it's capacity could satisfy the requirement.

No matter it's capacity smaller than requirement, or it's in state of uninitialized, it's both in state of unavailable for scheduler.

Reply:
I discussed with @saad-ali on this.  Our consensus is that it's better for topologyKey to be reported by the provisioner.  Depending on the user to specify the topologyKey, or even a flag, opens up the possibility that they could incorrectly set it.  And they don't have any choice in what to set as the value.  For local storage, it must always be set to "kubernetes.io/hostname", there is no other value that will work.

In the worst case, there will be a period of time during cluster initialization where the provisioners may not have reported capacity yet.  The consequences are not fatal.  Provisioning could fail until the provisioners are up and running (which is expected behavior in general).  Provisioning could succeed, and in that case we'll have to resync the capacity cache later.  We could try to mitigate this scenario, by recommending that provisioners do not process any provisioning requests until they have fully initialized (including reporting capacity).